torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1979,6 +1983,906 @@ def clip_grad_norm_(
+
+        Args:
+            model (torch.nn.Module): Root module (which may or may not be a
+                :class:`FullyShardedDataParallel` instance) whose parameters
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+            optim_input (Optional[Union[List[Dict[str, Any]],
+            Iterable[torch.nn.Parameter]]]): Input passed into the optimizer
+                ``optim`` representing either a :class:`list` of parameter
+                groups or an iterable of parameters; if ``None``, then this
+                method assumes the input was ``model.parameters()``. (Default:
+                ``None``)
+            target_rank (int): Rank on which to save the full optimizer state
+                dict in CPU memory. (Default: 0)
+
+        Returns:
+            full_osd (Dict[str, Any]): A :class:`dict` containing the optimizer
+                state for ``model`` 's original unflattened parameters and
+                having keys "state" and "param_groups" following the convention
+                of normal :class:`torch.optim.Optimizer` s.
+        """
+        FSDP = FullyShardedDataParallel  # alias
+        osd = optim.state_dict()
+        osd = FullyShardedDataParallel._preprocess_optim_state_dict(osd)
+        osd_state = osd["state"]  # alias
+        osd_param_groups = osd["param_groups"]  # alias
+
+        flat_param_to_fsdp_module = {}
+        for fsdp_module in FSDP.fsdp_modules(model):
+            for param in fsdp_module.params:  # may be empty
+                flat_param_to_fsdp_module[param] = fsdp_module
+
+        group = model.process_group if hasattr(model, "process_group") \
+            else None  # not all `torch.nn.Module`s have `process_group`
+        rank = dist.get_rank(group)
+        to_save = rank == target_rank
+        full_osd: Dict = {"state": {}, "param_groups": []} if to_save else {}
+        full_osd_state = full_osd["state"] if to_save else None  # alias
+
+        # Save a mapping from flattened to unflattened parameter IDs to remap
+        # the IDs in the "param_groups" part
+        flat_to_unflat_param_ids: Dict[int, List[int]] = {}
+        unflat_param_id = 0
+
+        # Handle the "state" part of the optimizer state dict
+        flat_param_id_to_param = \
+            FSDP._get_flat_param_id_to_param(model, optim_input)
+        for flat_param_id, param in enumerate(flat_param_id_to_param):  # type: ignore[assignment]
+            if flat_param_id not in flat_to_unflat_param_ids:
+                flat_to_unflat_param_ids[flat_param_id] = []
+            unflat_param_ids = flat_to_unflat_param_ids[flat_param_id]  # alias
+            # Do not include parameters without state to avoid empty mappings
+            if flat_param_id not in osd_state:
+                unflat_param_ids.append(unflat_param_id)
+                unflat_param_id += 1
+                continue
```

Comment:
If we have `linear1`, `linear2`, `linear3(bias=False)` in a single model but `linear2` is unused and hence has no optimizer state, then the mapping looks like:
```
# unflattened    flattened          notes
#           0            0    `linear1` weight
#           1            0    `linear1` bias
#           2         None    `linear2` weight  (no optim state)
#           3         None    `linear2` bias    (no optim state)
#           4            2    `linear3` weight
```
Thus, we see that we should still increment the `unflat_param_id`, or else `linear3.weight` will incorrectly map to `unflat_param_id` of `2`.

(This is the case where the parameter was passed into the optimizer constructor and hence is accounted for by the optimizer but it has not yet generated any optimizer state. It seems from Fairscale that this may arise for MoE.)